drivers/wifi/eswifi/eswifi_offload.c

```diff
@@ -0,0 +1,510 @@
+
+	/* copy payload */
+	offset = strlen(eswifi->buf);
+	for (frag = pkt->frags; frag; frag = frag->frags) {
+		memcpy(&eswifi->buf[offset], frag->data, frag->len);
+		offset += frag->len;
+	}
+
+	err = eswifi_request(eswifi, eswifi->buf, offset + 1,
+			     eswifi->buf, sizeof(eswifi->buf));
+	if (err || !__is_at_ok(eswifi->buf)) {
+		SYS_LOG_ERR("Unable to send data");
+		eswifi_unlock(eswifi);
+		return -EIO;
+	}
+
+	net_pkt_unref(pkt);
+	eswifi_unlock(eswifi);
+
+	return 0;
+}
+
+static int eswifi_off_sendto(struct net_pkt *pkt,
+			     const struct sockaddr *dst_addr,
+			     socklen_t addrlen,
+			     net_context_send_cb_t cb,
+			     s32_t timeout,
+			     void *token,
+			     void *user_data)
+{
+	/* TODO */
+	SYS_LOG_DBG("");
+	return -ENOTSUP;
+}
+
+static int eswifi_off_recv(struct net_context *context,
+			   net_context_recv_cb_t cb,
+			   s32_t timeout,
+			   void *user_data)
+{
+	struct eswifi_off_socket *socket = context->offload_context;
+	struct eswifi_dev *eswifi = eswifi_by_iface_idx(context->iface);
+
+	/* TODO */
+	SYS_LOG_DBG("");
+
+	eswifi_lock(eswifi);
+	socket->recv_cb = cb;
+	socket->user_data = user_data;
+	eswifi_unlock(eswifi);
+
+	return 0;
+}
+
+static int eswifi_off_put(struct net_context *context)
+{
+	struct eswifi_off_socket *socket = context->offload_context;
+	struct eswifi_dev *eswifi = eswifi_by_iface_idx(context->iface);
+	int err;
+
+	SYS_LOG_DBG("");
+
+	eswifi_lock(eswifi);
+
+	__select_socket(eswifi, socket->index);
+
+	socket->context = NULL;
+	socket->state = ESWIFI_SOCKET_STATE_NONE;
+
+	/* Stop TCP client */
+	sprintf(eswifi->buf, "P6=0\r");
+	err = eswifi_request(eswifi, eswifi->buf, strlen(eswifi->buf),
+			     eswifi->buf, sizeof(eswifi->buf));
+	if (err || !__is_at_ok(eswifi->buf)) {
+		SYS_LOG_ERR("Unable to connect");
+		return -EIO;
+	}
+
+	eswifi_unlock(eswifi);
+
+	return 0;
+}
+
+static struct net_offload eswifi_offload = {
+	.get	       = eswifi_off_get,
+	.bind	       = eswifi_off_bind,
+	.listen	       = eswifi_off_listen,
+	.connect       = eswifi_off_connect,
+	.accept	       = eswifi_off_accept,
+	.send	       = eswifi_off_send,
+	.sendto	       = eswifi_off_sendto,
+	.recv	       = eswifi_off_recv,
+	.put	       = eswifi_off_put,
+};
+
+static void eswifi_off_read_data(struct eswifi_dev *eswifi)
+{
+
+	int err, i;
+
+	eswifi_lock(eswifi);
+
+	for (i = 0; i < ESWIFI_OFFLOAD_MAX_SOCKETS; i++) {
+		struct eswifi_off_socket *socket = &eswifi->socket[i];
+		struct net_pkt *pkt;
+		struct net_buf *frag;
+
+		if (socket->state != ESWIFI_SOCKET_STATE_CONNECTED) {
+			continue;
+		}
+
+		SYS_LOG_DBG("");
+
+		err = __select_socket(eswifi, i);
+		if (err)
+			continue;
+
+		err = __read_data(eswifi, 1460); /* 1460 is max size */
+		if (err || !socket->recv_cb) {
+			continue;
+		}
+
+		pkt = net_pkt_get_reserve_rx(0, K_NO_WAIT);
+		if (!pkt) {
+			SYS_LOG_ERR("Cannot allocate rx packet");
+			continue;
+		}
+
+		frag = net_pkt_get_frag(pkt, K_NO_WAIT);
```

Comment:
I don't think you need to allocate a frag here, net_pkt_append_all() will do that for you if !pkt->frags.

Reply:
thanks good to know